java/code/src/com/redhat/rhn/domain/formula/FormulaFactory.java

```diff
@@ -1107,4 +1109,75 @@ public static String getClusterPillarDir() {
+        Map<String, Object> formulaValues = formulaData.getFormulaValues();
+        if (formulaValues.containsKey("exporters")) {
+            Optional<Boolean> proxyEnabled = getValueByPath(formulaValues, "proxy_enabled")
+                    .filter(Boolean.class::isInstance)
+                    .map(Boolean.class::cast);
+            Optional<Integer> proxyPort = Optional.empty();
+            if (proxyEnabled.isPresent() && proxyEnabled.get()) {
+                proxyPort = getValueByPath(formulaValues, "proxy_port")
+                        .filter(Number.class::isInstance)
+                        .map(Number.class::cast)
+                        .map(Number::intValue);
+            }
+
+            Map<String, Object> exportersMap = getValueByPath(formulaValues, "exporters")
+                    .filter(Map.class::isInstance)
+                    .map(Map.class::cast).get();
+
+            for (Map.Entry<String, Object> exporterEntry : exportersMap.entrySet()) {
+                Map<String, Object> exporterConfigMap = Optional.ofNullable(exporterEntry.getValue())
+                        .filter(Map.class::isInstance)
+                        .map(Map.class::cast).get();
+                Boolean exporterEnabled = Optional.ofNullable(exporterConfigMap.getOrDefault("enabled", false))
+                        .filter(Boolean.class::isInstance)
+                        .map(Boolean.class::cast).get();
+                if (exporterEnabled) {
+                    ExporterConfig exporterConfig = new ExporterConfig(
+                            exporterEntry.getKey(),
+                            Optional.ofNullable(exporterConfigMap.getOrDefault("address", null))
+                                    .filter(String.class::isInstance).map(String.class::cast).get(),
+                            Optional.ofNullable(exporterConfigMap.getOrDefault("args", null))
+                                    .filter(String.class::isInstance).map(String.class::cast).get());
+                    EndpointInfo endpointInfo = new EndpointInfo(
+                            formulaData.getSystemID(),
+                            Optional.ofNullable(exporterConfigMap.getOrDefault("name", exporterConfig.getName()))
+                                    .filter(String.class::isInstance).map(String.class::cast).get());
+                    endpointInfo.setExporterName(endpointInfo.getEndpointName());
+                    endpointInfo.setPort(proxyPort.orElse(exporterConfig.getPort()));
+                    if (proxyEnabled.isPresent() && proxyEnabled.get()) {
+                        endpointInfo.setModule(Optional.ofNullable(
+                                exporterConfigMap.getOrDefault("proxy_module", endpointInfo.getEndpointName()))
+                                .filter(String.class::isInstance).map(String.class::cast).get());
+                        endpointInfo.setPath("/proxy");
+                    }
+
+                    endpointInfos.add(endpointInfo);
```

Comment:
It would be cool if this algorithm could be refactored with streams, so that `EndpointInfo` could be turned into a read-only-after-construction class.

In general, we prefer creating collections via streams rather than modifying methods such as `put`, `add` and so on. Would that be possible here?